### DISCUSSION TITLE
Temperature and voltage in a multi Parameter Sweep Simulation

### DISCUSSION THREAD

**Discussion by bnilsson11** [ra3xdh/qucs_s#809] [Q&A] [answered]
I want to make a simulation of a SPICE Library Device, with two parameter sweeps.
One I-V, and the other with Temperature, to make several I-V graphs (T) in one chart.
It works in a single SW1 DC with temperature TEMP only and also with single SW1 DC voltage V1 only.
But when I use SW1 DC for voltage V1 and add SW2 SW1 for TEMP, temp becomes an unknown parameter in the sim log.
How should this be done?

![export](https://github.com/ra3xdh/qucs_s/assets/3372430/24fbd057-5a35-4f7e-a894-0d89df885c48)
[Constant_current_source_prj.zip](https://github.com/user-attachments/files/16113026/Constant_current_source_prj.zip)

**Comment by ra3xdh** [comment]
It is a limitation of the Ngspice, because Ngspice has no `.STEP` analysis like other simulators. The Parameter sweep is emulated by autogenrated script and this approach is very buggy and has several limitations. Only DC sweep can sweep temperature. The outer Parameter sweep loop is implemeter via `alter` or `alterparam` Ngspice command. These commands don't  accept global parameters like temperature.

**Comment by bnilsson11** [comment]
Is there any way I can set the temperature manually? I can compose the multigraph myself.  MvhBengt NilssonSkickat från min iPhone6 juli 2024 kl. 08:21 skrev Vadim Kuznetsov ***@***.***>:﻿
It is a limitation of the Ngspice, because Ngspice has no .STEP analysis like other simulators. The Parameter sweep is emulated by autogenrated script and this approach is very buggy and has several limitations. Only DC sweep can sweep temperature. The outer Parameter sweep loop is implemeter via alter or alterparam Ngspice command. These commands don't  accept global parameters like temperature.

**Comment by bnilsson11** [comment]
It worked to add a parameter Temp=85 directly in the SPICE lib file .model section.
To add Temp=TEMPER in the SPICE Lib .model and then an Equation TEMPER=85 did not work.
Setting Temp = 85 in the DC simulation dialog box did not work.
Is modifying the SPICE Lib file the only way?

**Comment by ra3xdh** [comment]
Refer to the section 1.3.2 of the Ngspice manual. https://ngspice.sourceforge.io/docs/ngspice-html-manual/manual.xhtml You can set the global temperature using `.OPTION`. This block could be found in *SPICE netlist sections*

![image](https://github.com/ra3xdh/qucs_s/assets/4920080/741f335a-0013-4863-9568-39103faa9f21)

**Comment by bnilsson11** [comment]
Perfect, thanks.

**Reply by tomhajjar** [reply]
Can you update your project above?

**Comment by bnilsson11** [comment]
> Can you update your project  above?

> 

I am sorry, I don't understand your request. Please clarify.

**Reply by tomhajjar** [reply]
Upload the corrected file

**Comment by bnilsson11** [comment]
Not much corrections, only the .PARAM added.
[Constant_current_source_prj.zip](https://github.com/user-attachments/files/16119470/Constant_current_source_prj.zip)

**Reply by tomhajjar** [reply]
I'm a bit confused. The "new" file has no .PARAM statement, it uses .OPTIONS and TEMP sweep was deleted.

![2024-07-07_095138](https://github.com/ra3xdh/qucs_s/assets/102838604/a8293fb4-fe92-41ed-8e59-e7e36fc9ff83)

**Reply by bnilsson11** [reply]
I am sorry, my bad. I wrote .PARAMS, should have been .OPTIONS.
I attach a project that sweeps TEMP.
As ra3xdh pointed out, a double sweep of V1 and TEMP does not work in this release.
Or did you really want the not working project?
[Constant_current_source_prj.zip](https://github.com/user-attachments/files/16119554/Constant_current_source_prj.zip)

**Reply by bnilsson11** [reply]
Maybe you were looking for this one?
[Temp_sensor_prj 2.zip](https://github.com/user-attachments/files/16119665/Temp_sensor_prj.2.zip)

**Comment by ra3xdh** [comment, marked as answer]
I have found a way how to automatize temperature sweep in outer loop. See #816 This feature will be available since the next release.

**Reply by bnilsson11** [reply]
That is wonderful news.
Next, release, meaning  24.2.2?

**Comment by dsm** [comment]
beginning of the 2024 I made this in qucs-s without a problem. (qucs-s 2.1.0) in parameter sweep I used TEMP keyword and in bc108b TEMPER keyword and simulation is done but newer version simulation failed.

![image](https://github.com/ra3xdh/qucs_s/assets/41967334/664c13b9-6bd7-4481-9d17-482bfacdc44a)

**Reply by dsm** [reply]
LM317 models different in qucs-s 2.1.0. in newer version LM317 models don't work but LM340_05 works.

**Reply by ra3xdh** [reply]
LM317 works on my machine as expected. The `TEMP` in Par.Sweep+DC and and in ParSweep+ParSweep works differently and the simulation failure is not related to this discussion. The schematic is failing because of other unknown reason.  Could you submit a schematic file?

![image](https://github.com/ra3xdh/qucs_s/assets/4920080/4109ab3f-5f34-4e85-a666-1497612f983c)

**Reply by dsm** [reply]
schematics is here. I replace the Lm317 to Lm340_05 everthing is work but old qucs-s ex. 2.1.0 Lm317 also work.

[temp_sensor_sim.zip](https://github.com/user-attachments/files/16119991/temp_sensor_sim.zip)

**Reply by ra3xdh** [reply]
I have tried temperatures sweep on LM317 circuits. The reason of the simulation failure is not related to Parameter sweep implementation itself. At some point an old LM317 model was replaced by true SPICE model from device vendor. An old model was from unknown origin and had numerous convergence issues. But it seems the new model is not designed to reflect the temperature behavior of the LM317 IC. It has a diode model having `N=0.01`. Usually N is in range from 1.0 to 2.0. This makes Ngspice failing to converge at some conditions.

**Reply by tomhajjar** [reply]
The LM317.cir I used to make the "original" library device is a TI "transient" Spice model. It required set ngbehavior=ltpsa to work. You changed it to be compatible without ngbehavior. For my testing I reverted back to the original model.

Diode D_D is connected from the output to ground and is normally reverse biased. Not sure why N=0.01. I changed it to 1.0 and the circuit works. 

VoltageRegulators.lib uses N=0.01 for the LM317 diode and N=.1 for the LM1084_ADJ diode. Both diodes are connected from the output to ground and are normally reverse biased. I see no reason not to make N= 1/2.

Problem has been around for a while. I posted the question over at ngspice.

https://ez.analog.com/design-tools-and-calculators/ltspice/f/q-a/572439/emission-coefficient-of-diode-model-in-spice-netlist

![2024-07-07_141121](https://github.com/ra3xdh/qucs_s/assets/102838604/d5a450ea-1661-48c0-a645-45cdebc80808)

**Comment by bobu01** [comment]
This is crude, but it generates the graph you wanted. 
![export](https://github.com/ra3xdh/qucs_s/assets/92656071/d57cb653-bdc9-4bc2-b653-017624a988f3)

[CCS_plot_cheat_prj.zip](https://github.com/user-attachments/files/16120369/CCS_plot_cheat_prj.zip)

**Reply by tomhajjar** [reply]
Not sure why devices rotated 90 deg and 1st device is missing it's symbol.

![2024-07-07_150410](https://github.com/ra3xdh/qucs_s/assets/102838604/4d4b33d0-d2a8-4370-bb9c-a10407e0f49f)

**Comment by tomhajjar** [comment]
@bobu01 
@ra3xdh 

Is the file above, [CCS_plot_cheat_prj.zip] corrupted? X1-X4 have issues on my Windows machine.

**Reply by bnilsson11** [reply]
I could run it on my Mac without issues, after I rotated the devices as needed.

**Reply by tomhajjar** [reply]
Understand but the question is why is the file "corrupted"? Making sure there isn't a "bug" effecting his PC.
